dali/pipeline/operator/direct_operator.h

```diff
@@ -0,0 +1,223 @@
+      const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs) {
+    DALI_FAIL("Unsupported backends in DirectOperator.Run().");
+  }
+
+  // Runs operator using specified thread pool.
+  template <typename InBackend, typename OutBackend>
+  DLL_PUBLIC std::vector<std::shared_ptr<TensorList<OutBackend>>> Run(
+      const std::vector<std::shared_ptr<TensorList<InBackend>>> &inputs,
+      const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs,
+      ThreadPool *tp) {
+    DALI_FAIL("Unsupported backends in DirectOperator.Run() with thread pool.");
+  }
+
+  // Runs operator using specified CUDA stream.
+  template <typename InBackend, typename OutBackend>
+  DLL_PUBLIC std::vector<std::shared_ptr<TensorList<OutBackend>>> Run(
+      const std::vector<std::shared_ptr<TensorList<InBackend>>> &inputs,
+      const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs,
+      cudaStream_t cuda_stream) {
+    DALI_FAIL("Unsupported backends in DirectOperator.Run() with CUDA stream");
+  }
+
+  DLL_PUBLIC inline static void SetThreadPool(int num_threads, int device_id, bool set_affinity) {
+    shared_thread_pool = std::make_unique<ThreadPool>(num_threads, device_id, set_affinity);
+  }
+
+  DLL_PUBLIC inline static void SetCudaStream(int device_id) {
+    if (device_id != CPU_ONLY_DEVICE_ID) {
+      DeviceGuard g(device_id);
+      shared_cuda_stream = CUDAStreamPool::instance().Get(device_id);
+    }
+  }
+
+ private:
+  template <typename InBackend, typename OutBackend, typename WSInputType, typename WSOutputType>
+  std::vector<std::shared_ptr<TensorList<OutBackend>>> RunImpl(
+      const std::vector<std::shared_ptr<TensorList<InBackend>>> &inputs,
+      const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs);
+
+  int batch_size;
+  size_t num_outputs;
+  workspace_t<Backend> ws;
+  std::unique_ptr<OperatorBase> op;
+
+  static cudaStream_t shared_cuda_stream;
+  static std::unique_ptr<ThreadPool> shared_thread_pool;
+};
+
+template <>
+template <>
+std::vector<std::shared_ptr<TensorList<CPUBackend>>> DirectOperator<CPUBackend>::Run(
+    const std::vector<std::shared_ptr<TensorList<CPUBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs,
+    ThreadPool *thread_pool) {
+  ws.SetThreadPool(thread_pool);
+
+  return RunImpl<CPUBackend, CPUBackend, TensorVector<CPUBackend>, TensorVector<CPUBackend>>(
+      inputs, kwargs);
+}
+
+template <>
+template <>
+std::vector<std::shared_ptr<TensorList<GPUBackend>>> DirectOperator<GPUBackend>::Run(
+    const std::vector<std::shared_ptr<TensorList<GPUBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs,
+    cudaStream_t cuda_stream) {
+  ws.set_stream(cuda_stream);
+  CUDA_CALL(cudaStreamSynchronize(cuda_stream));
```

Comment:
Long term we can make this synchronization optional.

Reply:
Why would we ever need to synchronize _before_?

Reply:
Removed the synchronization before the run and left only synchronization after. For now (especially in terms of debug mode) it shouldn't bother us but in the future it seems unnecessary to synchronize immediately after the run and only synchronize when we actually need this data (that synchronization before the run was kind of supposed to show that concept).